File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
import math
def init(arr, tree, node, start, end):
    if start == end:
        tree[node] = arr[start]
    else:
        init(arr,tree,node*2, start, (start+end)//2)
        init(arr,tree,node*2+1,(start+end)//2+1,end)
        tree[node] = tree[node*2] + tree[node*2+1]

def update(arr, tree, node, start, end, index, val):
    if index < start or index > end:
        return
    if start == end:
        arr[index] = val
        tree[node] = val
        return
    update(arr, tree, node*2, start, (start+end)//2, index, val)
    update(arr, tree, node*2+1, (start+end)//2+1, end, index, val)
    tree[node] = tree[node*2] + tree[node*2+1]

def query(tree,node,start,end,left,right):
    if left>end or right<start:
        return 0
    if left<=start and end<=right:
        return tree[node]
    lsum = query(tree,node*2,start,(start+end)//2,left,right)
    rsum = query(tree,node*2+1,(start+end)//2+1,end,left,right)
    return lsum + rsum


n, m, k  = map(int, input().split())
arr = [int(input()) for _ in range(n)]
h = math.ceil(math.log2(n))
size = 1<<(h+1)
tree = [0] * size
m += k
init(arr,tree,1,0,n-1)
for _ in range(m):
    command, t1, t2 = map(int,input().split())
    if command == 1:
        index, val = t1, t2
        update(arr,tree,1,0,n-1,index-1,val)
    else:
        left, right = t1, t2
        print(query(tree,1,0,n-1,left-1,right-1))

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br;
    static StringTokenizer st;
    static StringBuilder sb;
    static int n,m,k;
    static long [] tree;
    static long [] arr;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        sb = new StringBuilder();
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        int size = (int)Math.ceil(Math.log(n) / Math.log(2));
        tree = new long[2 << (size + 1)];
        arr = new long[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Long.parseLong(br.readLine());
        }
        m += k;
        init(1,0,n-1);
        while (m-- > 0) {
            String [] line = br.readLine().split(" ");
            int what = Integer.parseInt(line[0]);
            if (what == 1) {
                int index = Integer.parseInt(line[1]);
                long val = Long.parseLong(line[2]);
                update(1, 0, n-1, index-1, val);
            } else {
                int left = Integer.parseInt(line[1]);
                int right = Integer.parseInt(line[2]);
                sb.append(query( 1, 0, n-1, left-1, right-1)).append("\n");
            }
        }
        System.out.println(sb);
    }
    static void init(int node, int st, int end) {
        if (st == end) {
            tree[node] = arr[st];
        }
        else {
            init(node<<1, st, (st+end) /2 );
            init(node*2+1,(st+end)/2 + 1, end);
            tree[node] = tree[node<<1] + tree[node*2 + 1];
        }
    }
    static long query(int node, int st, int end, int l, int r) {
        if (l > end || r < st) return 0;
        if (l <= st && r >= end) return tree[node];
        long lsum = query(node<<1, st, (st+end) >> 1, l, r);
        long rsum= query(node*2 + 1, (st+end)/2 + 1, end, l, r);
        return lsum + rsum;
    }
    static void update(int node, int st, int end, int idx, long val) {
        if (idx < st || idx > end) return;
        if (st == end) {
            tree[node] = val;
            arr[idx] = val;
            return;
        }
        update(node<<1, st, (st+end) >> 1, idx, val);
        update(node*2 + 1, ((st+end) >> 1)+1, end, idx, val);
        tree[node] = tree[node<<1] + tree[node*2 + 1];
    }
}
